File: ts-interpreter/memory.ts
import { BoObject } from './object';

type map = Map<string, BoObject>;
export class Memory {
  super: Memory | undefined;
  slots: map;
  constructor(superMemory: Memory | undefined = undefined) {
    this.super = superMemory;
    this.slots = new Map<string, BoObject>();
  }
  subMemory() {
    return new Memory(this);
  }
  find(name: string): map | undefined {
    if (this.slots.get(name)) {
      return this.slots;
    } else if (this.super) {
      return this.super.find(name);
    } else {
      return undefined;
    }
  }
  get(name: string): BoObject | undefined {
    const slot = this.find(name);
    return slot && slot.get(name);
  }
  define(name: string, value: BoObject): BoObject | null {
    const isDefined = this.slots.get(name);
    if (!isDefined) {
      this.slots.set(name, value);
      return value;
    } else {
      return null;
    }
  }
  defineForce(name: string, value: BoObject): BoObject {
    this.slots.set(name, value);
    return value;
  }
  update(name: string, value: BoObject): BoObject | null {
    const slot = this.find(name);
    if (slot) {
      slot.set(name, value);
      return value;
    } else {
      return null;
    }
  }
  show() {
    this.super?.show();
    this.slots.forEach((value, name) => {
      console.log(`Slot.Show(${name} = ${value.str()})`);
    });
  }
}

File: go-compiler-vm/go-vm/memory.go
// memory.go — スコープチェーンによる変数管理。
package main

type Memory struct {
	Super *Memory
	Slots map[string]BoObject
	Order []string
}

func NewMemory(super *Memory) *Memory {
	return &Memory{
		Super: super,
		Slots: make(map[string]BoObject),
		Order: []string{},
	}
}

func (m *Memory) SubMemory() *Memory {
	return NewMemory(m)
}

func (m *Memory) Find(name string) map[string]BoObject {
	if _, ok := m.Slots[name]; ok {
		return m.Slots
	} else if m.Super != nil {
		return m.Super.Find(name)
	}
	return nil
}

func (m *Memory) Get(name string) BoObject {
	slot := m.Find(name)
	if slot != nil {
		return slot[name]
	}
	return nil
}

func (m *Memory) Define(name string, value BoObject) BoObject {
	if _, ok := m.Slots[name]; !ok {
		m.Slots[name] = value
		m.Order = append(m.Order, name)
		return value
	}
	return nil
}

func (m *Memory) DefineForce(name string, value BoObject) BoObject {
	if _, ok := m.Slots[name]; !ok {
		m.Order = append(m.Order, name)
	}
	m.Slots[name] = value
	return value
}

func (m *Memory) Update(name string, value BoObject) BoObject {
	slot := m.Find(name)
	if slot != nil {
		slot[name] = value
		return value
	}
	return nil
}
